src/OrleansRuntime/Core/Dispatcher.cs

```diff
@@ -636,9 +637,32 @@ private bool ShouldLogError(Exception ex)
         // Task returned by AsyncSendMessage()
         internal void SendMessage(Message message, ActivationData sendingActivation = null)
         {
+            if (TryAddressMessageFast(message))
+            {
+                TransportMessage(message);
+                return;
+            }
+
             AsyncSendMessage(message, sendingActivation).Ignore();
         }
 
+        private bool TryAddressMessageFast(Message message)
+        {
+            var targetAddress = message.TargetAddress;
+            if (targetAddress.IsComplete) return true;
+
+            AddressesAndTag addressesAndTag;
+
+            if (catalog.FastLookup(message.TargetGrain, out addressesAndTag))
+            {
+                // todo: consolidate with placement directors
```

Comment:
yes, you will need to consolidate it with placement. 
IPlacementRuntime has a method FastLookup.
you also need to carefully look at AddressMessage and make sure you use the same logic: for strategy, request and target.
and use SetTargetPlacement.